src/main/kotlin/com/team4099/robot2023/subsystems/drivetrain/swervemodule/threads/PhoenixOdometryThread.java

```diff
@@ -0,0 +1,110 @@
+ * a CANivore, the thread uses the "waitForAll" blocking method to enable more consistent sampling.
+ * This also allows Phoenix Pro users to benefit from lower latency between devices using CANivore
+ * time synchronization.
+ */
+public class PhoenixOdometryThread extends Thread {
+    private final Lock signalsLock =
+            new ReentrantLock(); // Prevents conflicts when registering signals
+    private BaseStatusSignal[] signals = new BaseStatusSignal[0];
+    private final List<Queue<Double>> queues = new ArrayList<>();
+    private boolean isCANFD = false;
+
+    private static PhoenixOdometryThread instance = null;
+
+    public static PhoenixOdometryThread getInstance() {
+        if (instance == null) {
+            instance = new PhoenixOdometryThread();
+        }
+        return instance;
+    }
+
+    private PhoenixOdometryThread() {
+        setName("PhoenixOdometryThread");
+        setDaemon(true);
+        start();
+    }
+
+    public Queue<Double> registerSignal(ParentDevice device, StatusSignal<Double> signal) {
+        isCANFD = Unmanaged.isNetworkFD(device.getNetwork());
+        Queue<Double> queue = new ArrayBlockingQueue<>(100);
+        signalsLock.lock();
+        try {
+            BaseStatusSignal[] newSignals = new BaseStatusSignal[signals.length + 1];
+            System.arraycopy(signals, 0, newSignals, 0, signals.length);
+            newSignals[signals.length] = signal;
+            signals = newSignals;
+            queues.add(queue);
+
+        } finally {
+            signalsLock.unlock();
+        }
+        return queue;
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            // Wait for updates from all signals
+            signalsLock.lock();
+            try {
+                if (isCANFD && signals.length > 0) {
+                    BaseStatusSignal.waitForAll(2.0 / DrivetrainConstants.OMOMETRY_UPDATE_FREQUENCY, signals);
+                } else {
+                    // "waitForAll" does not support blocking on multiple
+                    // signals with a bus that is not CAN FD, regardless
+                    // of Pro licensing. No reasoning for this behavior
+                    // is provided by the documentation.
+                    Thread.sleep((long) (1000.0 / DrivetrainConstants.OMOMETRY_UPDATE_FREQUENCY));
+                    BaseStatusSignal.refreshAll(signals);
+                }
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            } finally {
+                signalsLock.unlock();
+            }
+
+            // Save new data to queues
+            Drivetrain.Companion.setOdometryLock(true);
+            try {
+                for (int i = 0; i < signals.length; i++) {
+                    queues.get(i).offer(signals[i].getValueAsDouble());
+                }
+            } finally {
+                Drivetrain.Companion.setOdometryLock(true);
```

Comment:
no need to do companion, also this should be false. when you're done with saving data to queues, you wanna unlock

Reply:
is companion cause you need setOdometryLock to be static, otherwise you get error in cause its not static so you can't call the function without an object of the class